File: backend/src/application/use-cases/shopping-session/GetShoppingSessionMetricsUseCase.ts
import type { GetShoppingSessionMetricsUseCasePort } from "@application/ports/incoming/GetShoppingSessionMetricsUseCasePort";
import type { ShoppingSessionRepository } from "@application/ports/outgoing/ShoppingSessionRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type {
	GetShoppingSessionMetricsInput,
	GetShoppingSessionMetricsResult,
} from "./contracts";

export class GetShoppingSessionMetricsUseCase
	implements GetShoppingSessionMetricsUseCasePort
{
	constructor(
		private readonly shoppingSessionRepository: ShoppingSessionRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(
		input: GetShoppingSessionMetricsInput,
	): Promise<GetShoppingSessionMetricsResult> {
		this.logger.info(
			`[GetShoppingSessionMetricsUseCase] execute - userId: "${input.userId}"`,
		);

		if (input.userId.trim().length === 0) {
			throw new Error("User id is required to load shopping session metrics.");
		}

		const metrics = await this.shoppingSessionRepository.getMetricsForUser(
			input.userId,
		);

		return {
			supermarketDominance: metrics.supermarketDominance,
			spendingTrends: metrics.spendingTrends,
			ticketMetrics: metrics.ticketMetrics,
		};
	}
}

File: backend-dotnet/src/api/Application/UseCases/ShoppingSession/GetShoppingSessionMetricsUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;

namespace SupermarketPlanner.Api.Application.UseCases.ShoppingSession;

public sealed class GetShoppingSessionMetricsUseCase(
    IShoppingSessionRepository shoppingSessionRepository,
    ILogger<GetShoppingSessionMetricsUseCase> logger) : IGetShoppingSessionMetricsUseCase
{
    public Task<Either<DomainError, GetShoppingSessionMetricsResult>> Invoke(
        GetShoppingSessionMetricsInput input,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(input.UserId))
            return Task.FromResult(Either<DomainError, GetShoppingSessionMetricsResult>.FromLeft(
                DomainError.Validation(nameof(input.UserId), "User id is required to load shopping session metrics.")));

        logger.LogInformation(
            "[GetShoppingSessionMetricsUseCase] Invoke - userId: \"{UserId}\"",
            input.UserId);

        return shoppingSessionRepository.MetricsForUser(input.UserId, cancellationToken)
            .MapAsync(snapshot => new GetShoppingSessionMetricsResult(
                SupermarketDominance: snapshot.SupermarketDominance,
                SpendingTrends: snapshot.SpendingTrends,
                TicketMetrics: snapshot.TicketMetrics));
    }
}
